### ISSUE TITLE
GPU memory issues during fine-tuning FCN/CRF-RNN

### ISSUE THREAD

**Issue by AlexTS1980** [BVLC/caffe#5464]
### Issue summary

I'm trying to fine-tune FCNs and CRF-RNN to do an object tracking, and GPU memory has become a really big concern. I use a Tesla K40m with 12 Gb of memory and a workstation with 64Gb of RAM.

What I really want to know, if it is possible to predict how much GPU memory I'll need given the input parameters (image/batch size). I tried training with batch size=1 and original img size (1280x720), and got an out of memory issue. Then I reduced it to 1000x600: same. Finally, for 500x500 img it worked (reasonably fast) eating ~9.5 Gb of memory (same batch size). Obviously there's some correlation, and I don't know how to find it in order to predict the memory. 

What's even trickier, I don't quite understand how Caffe uses/reserves GPU memory, so I don't understand why it needs so much of it when processing just 1 image. I saved all layers from FCN, for 1280x720 image, that's 47 layers, 6.5Gb in total + weights 0.5Gb + diffs 0.5Gb = 7.5, far less than 12 Gb. So why does the error get reported? 

Summarizing, here are 2 questions: 1) Is there any way to predict GPU memory requirement given the input pars, 2) How does Caffe work with GPU memory? Are there any tricks to optimize/reduce the demand? 

PS I did ask the question on CUDA forums, not much help (https://devtalk.nvidia.com/default/topic/1000753/object-segmentation-with-cuda-memory-requirements)

Thanks

### Your system configuration
Operating system: Ubuntu 16.04
Compiler: C
CUDA version (if applicable): 8.0
CUDNN version (if applicable): 5.1
BLAS: version? I think python here uses openblas 3.2.18
Python or MATLAB version (for pycaffe and matcaffe respectively): Python

**Comment by williford** [comment]
Please ask your question on the [Caffe user mailing list](https://groups.google.com/forum/#!forum/caffe-users) or on another website, such as http://stackoverflow.com/.

> Please use the [caffe-users list](https://groups.google.com/forum/#!forum/caffe-users) for usage, installation, or modeling questions, or other requests for help.
> _Do not post such requests to Issues._ Doing so interferes with the development of Caffe.

**Comment by AlexTS1980** [comment]
I'm sorry I don't understand why this got closed. The question doesn't refer to the use, installation or bugs in Caffe. Memory mgmt is fundamental to caffe functioning and development.

**Comment by williford** [comment]
Make sure you take into account the size of calculating the gradients and momentum.

Caffe has a flood of Issues which are not related to the core development of Caffe. Because of this and not having a lot of resources, we (or they, as I'm not a "real" core member) have to be rather aggressive in closing Github Issues.

Caffe doesn't provide an easy way to calculate the memory usage. If you have improvements for memory usage or perhaps providing a way to calculate the memory usage, then a pull request or issue would be welcomed. In order to understand how much memory is being used, the mailing list or stackoverflow.com would be the preferred place to ask this. This is unfortunate since it would be nice to be able to discuss such things here.

I think it would be a good idea to ask on http://stackoverflow.com and link to the question on this issue.

I have also found that the K40 has more memory overhead than the Titan X. The difference is not huge, but there have been some models I have been able to train on the Titan X, but not on the K40.

See also (limited usefulness):
* https://groups.google.com/d/topic/caffe-users/BQSog9XaK2I/discussion
* https://groups.google.com/d/topic/caffe-users/ommgnUVXvrs/discussion
* https://github.com/BVLC/caffe/issues/4279
